File: app/src/test/java/io/github/hidroh/materialistic/data/FavoriteManagerTest.java
package io.github.hidroh.materialistic.data;

import android.accounts.Account;
import android.app.Notification;
import android.app.NotificationManager;
import android.arch.lifecycle.Observer;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.Uri;
import android.os.Parcel;
import android.preference.PreferenceManager;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.app.AppCompatActivity;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.Robolectric;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.annotation.Config;
import org.robolectric.shadows.ShadowContentResolver;
import org.robolectric.shadows.ShadowNetworkInfo;
import org.robolectric.util.ActivityController;

import java.io.File;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import javax.inject.Inject;

import io.github.hidroh.materialistic.BuildConfig;
import io.github.hidroh.materialistic.R;
import io.github.hidroh.materialistic.TestApplication;
import io.github.hidroh.materialistic.data.android.Cache;
import io.github.hidroh.materialistic.test.TestRunner;
import io.github.hidroh.materialistic.test.TestWebItem;
import io.github.hidroh.materialistic.test.shadow.ShadowWebView;
import rx.schedulers.Schedulers;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertTrue;
import static org.assertj.android.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.robolectric.Shadows.shadowOf;
import static org.robolectric.shadows.support.v4.Shadows.shadowOf;

@Config(shadows = {ShadowWebView.class})
@RunWith(TestRunner.class)
public class FavoriteManagerTest {
    @Inject MaterialisticDatabase.SavedStoriesDao savedStoriesDao;
    @Inject MaterialisticDatabase.ReadStoriesDao readStoriesDao;
    @Inject MaterialisticDatabase.ReadableDao readableDao;
    private ShadowContentResolver resolver;
    private FavoriteManager manager;
    @Mock Observer<Uri> observer;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        TestApplication.applicationGraph.inject(this);
        resolver = shadowOf(RuntimeEnvironment.application.getContentResolver());
        savedStoriesDao.insert(MaterialisticDatabase.SavedStory.from(new TestWebItem() {
            @Override
            public String getDisplayedTitle() {
                return "title";
            }

            @Override
            public String getUrl() {
                return "http://example.com";
            }

            @Override
            public String getId() {
                return "1";
            }
        }));
        savedStoriesDao.insert(MaterialisticDatabase.SavedStory.from(new TestWebItem() {
            @Override
            public String getDisplayedTitle() {
                return "ask HN";
            }

            @Override
            public String getUrl() {
                return "http://example.com";
            }

            @Override
            public String getId() {
                return "2";
            }
        }));
        LocalCache cache = new Cache(RuntimeEnvironment.application, savedStoriesDao, readStoriesDao, readableDao);
        manager = new FavoriteManager(cache, Schedulers.immediate(), savedStoriesDao) {
            @Override
            protected Uri getUriForFile(Context context, File file) {
                return Uri.parse("content://" + FavoriteManager.FILE_AUTHORITY + "/files/saved/materialistic-export.txt");
            }
        };
        MaterialisticDatabase.getInstance(RuntimeEnvironment.application).getLiveData().observeForever(observer);
        reset(observer);
    }

    @Test
    public void testLocalItemManager() {
        ActivityController<AppCompatActivity> controller = Robolectric.buildActivity(AppCompatActivity.class);
        AppCompatActivity activity = controller.create().start().resume().get();
        LocalItemManager.Observer observer = mock(LocalItemManager.Observer.class);
        manager.attach(observer, null);
        verify(observer).onChanged();
        assertThat(manager.getSize()).isEqualTo(2);
        assertThat(manager.getItem(0).getDisplayedTitle()).contains("ask HN");
        assertThat(manager.getItem(1).getDisplayedTitle()).contains("title");
        manager.detach();
        controller.pause().stop().destroy();
    }

    @Test
    public void testExportNoQuery() {
        manager.export(RuntimeEnvironment.application, null);
        List<Notification> allNotifications = shadowOf((NotificationManager)
                RuntimeEnvironment.application
                        .getSystemService(Context.NOTIFICATION_SERVICE))
                .getAllNotifications();
        assertThat(allNotifications).isNotEmpty();
        assertThat(shadowOf(allNotifications.get(0).contentIntent).getSavedIntent())
                .hasAction(Intent.ACTION_CHOOSER);
    }

    @Test
    public void testExportEmpty() {
        manager.export(RuntimeEnvironment.application, "blah");
        assertThat(shadowOf((NotificationManager) RuntimeEnvironment.application
                .getSystemService(Context.NOTIFICATION_SERVICE))
                .getAllNotifications())
                .isEmpty();
    }

    @Test
    public void testCheckNoId() {
        assertFalse(manager.check(null)
                .toBlocking().single());
    }

    @Test
    public void testCheckTrue() {
        assertTrue(manager.check("1")
                .toBlocking().single());
    }

    @Test
    public void testCheckFalse() {
        assertFalse(manager.check("-1")
                .toBlocking().single());
    }

    @Config(sdk = 18)
    @Test
    public void testAdd() {
        PreferenceManager.getDefaultSharedPreferences(RuntimeEnvironment.application)
                .edit()
                .putBoolean(RuntimeEnvironment.application
                        .getString(R.string.pref_saved_item_sync), true)
                .putBoolean(RuntimeEnvironment.application
                        .getString(R.string.pref_offline_article), true)
                .apply();
        shadowOf((ConnectivityManager) RuntimeEnvironment.application
                .getSystemService(Context.CONNECTIVITY_SERVICE))
                .setActiveNetworkInfo(ShadowNetworkInfo.newInstance(null,
                        ConnectivityManager.TYPE_WIFI, 0, true, true));
        manager.add(RuntimeEnvironment.application, new TestWebItem() {
            @Override
            public String getId() {
                return "3";
            }

            @Override
            public String getUrl() {
                return "http://newitem.com";
            }

            @Override
            public String getDisplayedTitle() {
                return "new title";
            }
        });
        verify(observer).onChanged(any(Uri.class));
        assertTrue(ShadowContentResolver.isSyncActive(new Account("Materialistic", BuildConfig.APPLICATION_ID),
                SyncContentProvider.PROVIDER_AUTHORITY));
    }

    @Test
    public void testReAdd() {
        Favorite favorite = mock(Favorite.class);
        when(favorite.getId()).thenReturn("3");
        when(favorite.getUrl()).thenReturn("http://example.com");
        when(favorite.getDisplayedTitle()).thenReturn("title");
        manager.add(RuntimeEnvironment.application, favorite);
        verify(observer).onChanged(any(Uri.class));
    }

    @Test
    public void testClearAll() {
        manager.clear(RuntimeEnvironment.application, null);
        verify(observer).onChanged(any(Uri.class));
    }

    @Test
    public void testClearQuery() {
        manager.clear(RuntimeEnvironment.application, "blah");
        verify(observer).onChanged(any(Uri.class));
    }

    @Test
    public void testRemoveNoId() {
        manager.remove(RuntimeEnvironment.application, (String) null);
        assertThat(shadowOf(LocalBroadcastManager.getInstance(RuntimeEnvironment.application))
                .getSentBroadcastIntents()).isEmpty();
    }

    @Test
    public void testRemoveId() {
        manager.remove(RuntimeEnvironment.application, "1");
        verify(observer).onChanged(any(Uri.class));
    }

    @Test
    public void testRemoveMultipleNoId() {
        manager.remove(RuntimeEnvironment.application, (Set<String>) null);
        verify(observer, never()).onChanged(any(Uri.class));
    }

    @Test
    public void testRemoveMultiple() {
        manager.remove(RuntimeEnvironment.application, new HashSet<String>(){{add("1");add("2");}});
        verify(observer, times(2)).onChanged(any(Uri.class));
    }

    @Test
    public void testFavorite() {
        Parcel parcel = Parcel.obtain();
        parcel.writeString("1");
        parcel.writeString("http://example.com");
        parcel.writeString("title");
        parcel.setDataPosition(0);
        Favorite favorite = Favorite.CREATOR.createFromParcel(parcel);
        assertEquals("title", favorite.getDisplayedTitle());
        assertEquals("example.com", favorite.getSource());
        assertEquals("http://example.com", favorite.getUrl());
        assertEquals("1", favorite.getId());
        assertNotNull(favorite.getDisplayedAuthor(RuntimeEnvironment.application, true, 0));
        assertEquals(Item.STORY_TYPE, favorite.getType());
        assertTrue(favorite.isStoryType());
        assertEquals("title (http://example.com) - https://news.ycombinator.com/item?id=1", favorite.toString());
        assertEquals(0, favorite.describeContents());
        Parcel output = Parcel.obtain();
        favorite.writeToParcel(output, 0);
        output.setDataPosition(0);
        assertEquals("1", output.readString());
        assertThat(Favorite.CREATOR.newArray(1)).hasSize(1);
    }

    @After
    public void tearDown() {
        MaterialisticDatabase.getInstance(RuntimeEnvironment.application).getLiveData().removeObserver(observer);
    }
}

File: app/src/test/java/io/github/hidroh/materialistic/data/FavoriteManagerTest.kt
/*
 * Copyright (c) 2018 Ha Duy Trung
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.hidroh.materialistic.data

import android.accounts.Account
import android.app.NotificationManager
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.net.NetworkInfo
import android.net.Uri
import android.preference.PreferenceManager
import androidx.lifecycle.Observer
import io.github.hidroh.materialistic.BuildConfig
import io.github.hidroh.materialistic.R
import io.github.hidroh.materialistic.data.android.Cache
import io.github.hidroh.materialistic.test.InMemoryDatabase
import io.github.hidroh.materialistic.test.TestRunner
import io.github.hidroh.materialistic.test.TestWebItem
import org.assertj.core.api.Assertions.assertThat
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.MockitoAnnotations
import org.robolectric.RuntimeEnvironment
import org.robolectric.Shadows.shadowOf
import org.robolectric.annotation.Config
import org.robolectric.shadows.ShadowContentResolver
import org.robolectric.shadows.ShadowNetworkInfo
import rx.schedulers.Schedulers

@RunWith(TestRunner::class)
class FavoriteManagerTest {
  private val context = RuntimeEnvironment.application
  private val database = InMemoryDatabase.getInstance(RuntimeEnvironment.application)
  private val savedStoriesDao = database.savedStoriesDao
  private val readStoriesDao = database.readStoriesDao
  private val readableDao = database.readableDao
  private lateinit var manager: FavoriteManager
  @Mock private lateinit var observer: Observer<Uri>

  @Before
  fun setUp() {
    MockitoAnnotations.initMocks(this)
    savedStoriesDao.insert(MaterialisticDatabase.SavedStory.from(object : TestWebItem() {
      override fun getDisplayedTitle() = "title"
      override fun getUrl() = "http://example.com"
      override fun getId() = "1"
    }))
    savedStoriesDao.insert(MaterialisticDatabase.SavedStory.from(object : TestWebItem() {
      override fun getDisplayedTitle() = "ask HN"
      override fun getUrl() = "http://example.com"
      override fun getId() = "2"
    }))
    val cache = Cache(database, savedStoriesDao, readStoriesDao, readableDao, Schedulers.immediate())
    manager = FavoriteManager(cache, Schedulers.immediate(), savedStoriesDao)
    MaterialisticDatabase.getInstance(context).liveData.observeForever(observer)
  }

  @Test
  fun testLocalItemManager() {
    val observer = mock(LocalItemManager.Observer::class.java)
    manager.attach(observer, null)
    verify(observer).onChanged()
    assertThat(manager.size).isEqualTo(2)
    assertThat(manager.getItem(0)!!.displayedTitle).contains("ask HN")
    assertThat(manager.getItem(1)!!.displayedTitle).contains("title")
    manager.detach()
  }

  @Test
  fun testExportNoQuery() {
    manager.export(context, null)
    val notifications = shadowOf(context
        .getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager)
        .allNotifications
    assertThat(notifications).isNotEmpty()
    assertThat(shadowOf(notifications[0].contentIntent).savedIntent.action).isEqualTo(Intent.ACTION_CHOOSER)
  }

  @Test
  fun testExportEmpty() {
    manager.export(context, "blah")
    assertThat(shadowOf(context
        .getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager)
        .allNotifications)
        .isEmpty()
  }

  @Test
  fun testCheckNoId() {
    assertThat(manager.check(null).toBlocking().single()).isFalse()
  }

  @Test
  fun testCheckTrue() {
    assertThat(manager.check("1").toBlocking().single()).isTrue()
  }

  @Test
  fun testCheckFalse() {
    assertThat(manager.check("-1").toBlocking().single()).isFalse()
  }

  @Test @Config(sdk = [18])
  fun testAdd() {
    PreferenceManager.getDefaultSharedPreferences(context)
        .edit()
        .putBoolean(context.getString(R.string.pref_saved_item_sync), true)
        .putBoolean(context.getString(R.string.pref_offline_article), true)
        .apply()
    shadowOf(context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager)
        .setActiveNetworkInfo(ShadowNetworkInfo.newInstance(null,
        ConnectivityManager.TYPE_WIFI, 0, true, NetworkInfo.State.CONNECTED))
    manager.add(context, object : TestWebItem() {
      override fun getDisplayedTitle() = "new title"
      override fun getUrl() = "http://newitem.com"
      override fun getId() = "3"
    })
    verify(observer).onChanged(eq(Uri.parse("content://${BuildConfig.APPLICATION_ID}/saved/add/3")))
    assertThat(ShadowContentResolver.getStatus(Account("Materialistic", BuildConfig.APPLICATION_ID),
        SyncContentProvider.PROVIDER_AUTHORITY).syncRequests).isGreaterThan(0)
  }

  @Test
  fun testReAdd() {
    val favorite = mock(Favorite::class.java)
    `when`(favorite.id).thenReturn("3")
    `when`(favorite.url).thenReturn("http://example.com")
    `when`(favorite.displayedTitle).thenReturn("title")
    manager.add(context, favorite)
    verify(observer).onChanged(eq(Uri.parse("content://${BuildConfig.APPLICATION_ID}/saved/add/3")))
  }

  @Test
  fun testClearAll() {
    manager.clear(context, null)
    verify(observer).onChanged(eq(Uri.parse("content://${BuildConfig.APPLICATION_ID}/saved/clear")))
  }

  @Test
  fun testClearQuery() {
    manager.clear(context, "blah")
    verify(observer).onChanged(eq(Uri.parse("content://${BuildConfig.APPLICATION_ID}/saved/clear")))
  }

  @Test
  fun testRemoveWithId() {
    manager.remove(context, "1")
    verify(observer).onChanged(eq(Uri.parse("content://${BuildConfig.APPLICATION_ID}/saved/remove/1")))
  }

  @Test
  fun testRemoveMultipleNoId() {
    manager.remove(context, null as Set<String>?)
    verify(observer, never()).onChanged(any(Uri::class.java))
  }

  @Test
  fun testRemoveMultipleWithId() {
    manager.remove(context, hashSetOf("1", "2"))
    verify(observer).onChanged(eq(Uri.parse("content://${BuildConfig.APPLICATION_ID}/saved/remove/1")))
    verify(observer).onChanged(eq(Uri.parse("content://${BuildConfig.APPLICATION_ID}/saved/remove/2")))
  }

  @After
  fun tearDown() {
    MaterialisticDatabase.getInstance(context).liveData.removeObserver(observer)
  }
}
